compiler/rustc_ty_utils/src/layout_sanity_check.rs

```diff
@@ -296,4 +297,50 @@ pub(super) fn sanity_check_layout<'tcx>(
             }
         }
     }
+
+    // If the struct tail is sized and can be unsized, check that unsizing doesn't move the fields around.
+    if let ty::Adt(def, substs) = *layout.ty.kind()
+        && def.is_struct()
+        && let Some(tail_def) = def.variant(FIRST_VARIANT).fields.raw.last()
+        && let tail_binder = cx.tcx.type_of(tail_def.did)
+        && tail_binder.subst(cx.tcx, substs).is_sized(cx.tcx, cx.param_env)
+        && !tail_binder.subst_identity().is_sized(cx.tcx, cx.tcx.param_env(def.did()))
+        && let ty::Param(tail_param) = tail_binder.subst_identity().kind()
```

Comment:
Apparently we also support support unsizing if the tail is a projection and the current implementation cannot reasonably handle this case.

```rust
trait Tr {
    type Assoc: ?Sized;
}

impl Tr for i32 {
    type Assoc = [(); 1];
}

impl Tr for i64 {
    type Assoc = [()];
}

struct S<T: Tr + ?Sized> {
    tail: <T as Tr>::Assoc,
}

fn main() {
    let x: &S<i64> = &S::<i32> {
        tail: [()],
    };
}
```